apis/bigqueryconnection/v1beta1/connection_reference.go

```diff
@@ -100,22 +86,35 @@ type BigQueryConnectionConnectionRef struct {
 	Name string `json:"name,omitempty"`
 	// The `namespace` of a `BigQueryConnectionConnection` resource.
 	Namespace string `json:"namespace,omitempty"`
+}
 
-	parent string
+func parseExternal(external string) (string, string, string, error) {
+	tokens := strings.Split(external, "/")
+	if len(tokens) != 6 || tokens[0] != "projects" || tokens[2] != "locations" || tokens[4] != "connections" {
+		return "", "", "", fmt.Errorf("external should be projects/<project>/locations/<location>/connections/<Connection>, got %s", external)
+	}
+	return tokens[1], tokens[3], tokens[5], nil
 }
 
 func (r *BigQueryConnectionConnectionRef) Parent() (string, error) {
-	if r.parent != "" {
-		return r.parent, nil
-	}
 	if r.External != "" {
 		r.External = strings.TrimPrefix(r.External, "/")
-		tokens := strings.Split(r.External, "/")
-		if len(tokens) != 6 || tokens[0] != "projects" || tokens[2] != "locations" || tokens[4] != "connections" {
-			return "", fmt.Errorf("format of BigQueryConnectionConnection external=%q was not known (use projects/<projectId>/locations/<location>/connections/<connectionID>)", r.External)
+		project, location, _, err := parseExternal(r.External)
+		if err != nil {
+			return "", err
 		}
-		r.parent = "projects/" + tokens[1] + "/locations/" + tokens[3]
-		return r.parent, nil
+		return "projects/" + project + "/locations/" + location, nil
+	}
+	return "", fmt.Errorf("BigQueryConnectionConnectionRef not normalized to External form or not created from `New()`")
+}
+
+func (r *BigQueryConnectionConnectionRef) ConnectionID() (string, error) {
+	if r.External != "" {
+		_, _, id, err := parseExternal(r.External)
```

Comment:
I see. this can return empty id. Can we add a comment here? I feel like this is a special case that we should reflect in the code. If not, we'd better add a comment to call it out.

for example
```go
type ConnectionID struct {
      parent string 
      id         string 
      unspecified bool
}

func parseExternal() (*connectionID, error) {}

func Find() {
	connectionID, error := parseExternal()
        if connectionID.unspecified {
             return false, nil
        }
}
```

Reply:
Thanks!